### PR TITLE
Don't collapse dimensions during batched matmul (FIX #799)

### LINKED ISSUE #799: Broadcasting discrepancies
```python
import torch
from tinygrad.tensor import Tensor

A = torch.ones(4, 3)
B = torch.ones(1, 3, 3, 3)
C = A @ B


# torch torch.Size([1, 3, 4, 3])
print('torch', C.shape)

D = Tensor.ones(4, 3)
E = Tensor.ones(1, 3, 3, 3)
F = D @ E

# tinygrad (4, 9)
print('tinygrad', F.shape)
```

ref: 
https://pytorch.org/docs/stable/notes/broadcasting.html
https://numpy.org/doc/stable/user/basics.broadcasting.html
https://github.com/geohot/tinygrad/pull/114

### PR DESCRIPTION
**Pull request by aneeshdurg**
The reshape should only be necessary when multiplying a vector with a matrix - this fixes batched matmul.

### CODE REVIEW

File: tinygrad/tensor.py [hunk 1]
```diff
@@ -388,7 +388,8 @@ def conv2d(self, weight:Tensor, bias:Optional[Tensor]=None, groups=1, stride=1,
   def dot(self, w:Tensor) -> Tensor:
     x = self.reshape(*self.shape[0:-1], 1, self.shape[-1])
     w = w.reshape(*w.shape[0:-2], 1, w.shape[-2], w.shape[-1]).transpose(-1, -2)
-    return (x*w).sum(-1).reshape(*x.shape[0:-2], -1)
+    r = (x*w).sum(-1)
+    return r.reshape((*r.shape[:-2], r.shape[-1]) if len(self.shape) == 1 else r.shape)
```
Comment (geohot) [inline tinygrad/tensor.py:392]: `r if len(self.shape) == 1 else r.reshape((*r.shape[:-2], r.shape[-1]))` is the same? Avoid the r.reshape(r.shape)

**Comment by geohot** [comment]
Also, tests failing.

**Comment by aneeshdurg** [comment]
Looking at the failing tests, it looks like the values are still pretty close to the expected values...I'm unable to repro that locally, but I'll keep digging

**Comment by geohot** [comment]
The OpenCL test failure is shape related.

**Comment by trunghd113** [comment]
nice

**Comment by geohot** [comment]
Hmm, broadcastdot works okay on `IMAGE`, what's the issue? Raising exceptions is fine, `RuntimeError`

**Comment by geohot** [comment]
Will merge for now, thanks!

File: tinygrad/tensor.py [hunk 2]
```diff
@@ -388,7 +388,8 @@ def conv2d(self, weight:Tensor, bias:Optional[Tensor]=None, groups=1, stride=1,
   def dot(self, w:Tensor) -> Tensor:
     x = self.reshape(*self.shape[0:-1], 1, self.shape[-1])
     w = w.reshape(*w.shape[0:-2], 1, w.shape[-2], w.shape[-1]).transpose(-1, -2)
-    return (x*w).sum(-1).reshape(*x.shape[0:-2], -1)
+    r = (x*w).sum(-1)
+    return r.reshape((*r.shape[:-2], r.shape[-1])) if len(self.shape) == 1 else r
```
Comment (geohot) [inline tinygrad/tensor.py:392]: NIT: you don't need the extra parens here.

**Pull request merged**